File: Akshit_22b2531/week2/sum_of_three_values.py
def solve():
    n, x = map(int, input().split())
    arr = list(map(int, input().split()))
    
    # Store original indices
    indexed_arr = [(arr[i], i+1) for i in range(n)]
    
    # Sort by value
    indexed_arr.sort()
    
    # For each element, try to find two other elements that sum to x - element
    for i in range(n):
        target = x - indexed_arr[i][0]
        
        # Use two pointers to find the pair
        left = 0
        right = n - 1
        
        while left < right:
            if left == i:
                left += 1
                continue
            if right == i:
                right -= 1
                continue
                
            current_sum = indexed_arr[left][0] + indexed_arr[right][0]
            
            if current_sum == target:
                # Found a solution
                return f"{indexed_arr[i][1]} {indexed_arr[left][1]} {indexed_arr[right][1]}"
            elif current_sum < target:
                left += 1
            else:
                right -= 1
    
    return "IMPOSSIBLE"

if __name__ == "__main__":
    print(solve()) 

File: week2/Sum of Three Values.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    long long x;
    cin >> n >> x;

    vector<pair<long long, int>> arr(n);
    for (int i = 0; i < n; ++i) {
        cin >> arr[i].first;
        arr[i].second = i + 1;
    }
    sort(arr.begin(), arr.end());
    for (int i = 0; i < n; ++i) {
        int left = i + 1;
        int right = n - 1;
        long long target = x - arr[i].first;

        while (left < right) {
            long long sum = arr[left].first + arr[right].first;
            if (sum == target) {
                cout << arr[i].second << " " << arr[left].second << " " << arr[right].second << "\n";
                return 0;
            }
            if (sum < target)
                ++left;
            else
                --right;
        }
    }
    cout << "IMPOSSIBLE\n";
    return 0;
}
